stdlib/ephemeron.mli

```diff
@@ -0,0 +1,306 @@
+  val blit_key12 : ('k1,'k2,_) t -> ('k1,'k2,_) t -> unit
+  (** Same as {!Ephemeron.K1.blit_key} *)
+
+  val get_data: ('k1,'k2,'d) t -> 'd option
+  (** Same as {!Ephemeron.K1.get_data} *)
+  val get_data_copy: ('k1,'k2,'d) t -> 'd option
+  (** Same as {!Ephemeron.K1.get_data_copy} *)
+  val set_data: ('k1,'k2,'d) t -> 'd -> unit
+  (** Same as {!Ephemeron.K1.set_data} *)
+  val unset_data: ('k1,'k2,'d) t -> unit
+  (** Same as {!Ephemeron.K1.unset_data} *)
+  val check_data: ('k1,'k2,'d) t -> bool
+  (** Same as {!Ephemeron.K1.check_data} *)
+  val blit_data: ('k1,'k2,'d) t -> ('k1,'k2,'d) t -> unit
+  (** Same as {!Ephemeron.K1.blit_data} *)
+
+  module Make
+      (H1:Hashtbl.HashedType)
+      (H2:Hashtbl.HashedType) :
+    S with type key = H1.t * H2.t
+  (** Functor building an implementation of a weak hash table *)
+
+  module MakeSeeded
+      (H1:Hashtbl.SeededHashedType)
+      (H2:Hashtbl.SeededHashedType) :
+    SeededS with type key = H1.t * H2.t
+  (** Functor building an implementation of a weak hash table.
+      The seed is similar to the one of {!Hashtbl.MakeSeeded}. *)
+
+end
+
+module Kn : sig
+  type ('k,'d) t (** an ephemeron with an arbitrary number of keys
+                      of the same types *)
+
+  val create: int -> ('k,'d) t
+  (** Same as {!Ephemeron.K1.create} *)
+
+  val get_key: ('k,'d) t -> int -> 'k option
+  (** Same as {!Ephemeron.K1.get_key} *)
+  val get_key_copy: ('k,'d) t -> int -> 'k option
+  (** Same as {!Ephemeron.K1.get_key_copy} *)
+  val set_key: ('k,'d) t -> int -> 'k -> unit
+  (** Same as {!Ephemeron.K1.set_key} *)
+  val unset_key: ('k,'d) t -> int -> unit
+  (** Same as {!Ephemeron.K1.unset_key} *)
+  val check_key: ('k,'d) t -> int ->  bool
+  (** Same as {!Ephemeron.K1.check_key} *)
+
+  val blit_key  : ('k,_) t -> int -> ('k,_) t -> int -> int -> unit
+  (** Same as {!Ephemeron.K1.blit_key} *)
+
+  val get_data: ('k,'d) t -> 'd option
+  (** Same as {!Ephemeron.K1.get_data} *)
+  val get_data_copy: ('k,'d) t -> 'd option
+  (** Same as {!Ephemeron.K1.get_data_copy} *)
+  val set_data: ('k,'d) t -> 'd -> unit
+  (** Same as {!Ephemeron.K1.set_data} *)
+  val unset_data: ('k,'d) t -> unit
+  (** Same as {!Ephemeron.K1.unset_data} *)
+  val check_data: ('k,'d) t -> bool
+  (** Same as {!Ephemeron.K1.check_data} *)
+  val blit_data: ('k,'d) t -> ('k,'d) t -> unit
+  (** Same as {!Ephemeron.K1.blit_data} *)
+
+  module Make
+      (H:Hashtbl.HashedType) :
+    S with type key = H.t array
+  (** Functor building an implementation of a weak hash table *)
+
+  module MakeSeeded
+      (H:Hashtbl.SeededHashedType) :
+    SeededS with type key = H.t array
+  (** Functor building an implementation of a weak hash table.
+      The seed is similar to the one of {!Hashtbl.MakeSeeded}. *)
+
+end
+
+module GenHashTable: sig
+  (** Define hash table on generic containers. *)
+
+  (** It can be used in conjunction of {!Obj.Ephemeron}
+      for building weak hash table for specific type of keys *)
```

Comment:
I'm quite uncomfortable with the idea of advertising the use of the `Obj` module. It may mean that a preinstantiated version of this functor could be exported if there is a safe way to do that.